Fix object creation performance regression #1192

zalokhan opened the pull request:
## Overview

Description:
Fixing performance. Instead of creating deep copies of the layout, reusing the layout reference in the RuntimeLayout constructor.
In the AbstractView, we now keep an atomic reference of the RuntimeLayout used and create new only if there is an epoch change. 

Why should this be merged: Fixes performance drop introduced by #1192 

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

Maithem commented on runtime/src/main/java/org/corfudb/runtime/view/RuntimeLayout.java:
```diff
@@ -49,7 +49,7 @@
     public RuntimeLayout(@Nonnull Layout layout, @Nonnull CorfuRuntime corfuRuntime) {
 
-        this.layout = new Layout(layout);
+        this.layout = layout;
```
I thought the whole point of a runtime layout is to have your own layout.

zalokhan replied:
It still does right. It's just the management agent which modifies layouts. We can make sure all modifications are made on a copy.

no2chem replied:
Yeah, this is fine. a runtime layout is a runtime+layout so you can easily get new clients.

no2chem commented on runtime/src/main/java/org/corfudb/runtime/view/AbstractView.java:
```diff
@@ -59,6 +61,7 @@ public Layout getCurrentLayout() {
         return layoutHelper(function, false);
     }
 
+    private volatile AtomicReference<RuntimeLayout> runtimeLayout = new AtomicReference<>();
```
I'm pretty sure this doesn't need to be an AtomicReference.

medhavidhawan replied:
remove volatile

no2chem commented on runtime/src/main/java/org/corfudb/runtime/view/AbstractView.java:
```diff
@@ -86,7 +89,13 @@ public Layout getCurrentLayout() {
         while (true) {
             try {
-                return function.apply(new RuntimeLayout(runtime.layout.get(), runtime));
+                final Layout layout = runtime.layout.get();
```
Place proper synchronization around this so only a single retrieval is permitted, use a stamped lock so that optimistic reads are possible.

zalokhan replied:
I used the atomic reference to provide synchronization. So that single retrieval is permitted.

corfudb-performance commented:
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **a85bc12**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1229 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/nVoMByk2S4OWaaZPA26XLOwmDhXbKKhP)

medhavidhawan commented on runtime/src/main/java/org/corfudb/runtime/view/RuntimeLayout.java:
```diff
@@ -49,7 +49,7 @@
     public RuntimeLayout(@Nonnull Layout layout, @Nonnull CorfuRuntime corfuRuntime) {
 
-        this.layout = new Layout(layout);
+        this.layout = layout;
```
change the old comment.

zalokhan commented:
@no2chem @medhavidhawan I have addressed the comments.
Michael, we agreed on AtomicReference.
I have confirmed with Anny that the performance numbers have improved matching the commits before the regression.
